lsp/lsp.cpp

```diff
@@ -73,265 +73,281 @@ static const char *MethodNotFound       = "-32601";
+
+      while (true) {
+        size_t json_size = 0;
+        // Read header lines until an empty line
+        while (true) {
+          std::string line;
+          std::getline(std::cin, line);
+          // Trim trailing CR, if any
+          if (!line.empty() && line.back() == '\r')
+            line.resize(line.size() - 1);
+          // EOF? exit LSP
+          if (std::cin.eof())
+            exit(0);
+          // Failure reading? Fail with non-zero exit status
+          if (!std::cin)
+            exit(1);
+          // Empty line? stop
+          if (line.empty())
+            break;
+          // Capture the json_size
+          if (line.compare(0, sizeof(contentLength) - 1, &contentLength[0]) == 0)
+            json_size = std::stoul(line.substr(sizeof(contentLength) - 1));
+        }
 
-JAST createResponseMessage() {
-  JAST message = createMessage();
-  message.add("id", JSON_NULLVAL);
-  return message;
-}
+        // Content-Length was unset?
+        if (json_size == 0)
+          exit(1);
 
-JAST createResponseMessage(JAST receivedMessage) {
-  JAST message = createMessage();
-  message.children.emplace_back("id", receivedMessage.get("id"));
-  return message;
-}
+        // Retrieve the content
+        std::string content(json_size, ' ');
+        std::cin.read(&content[0], json_size);
 
-void sendErrorMessage(const char *code, std::string message) {
-  JAST errorMessage = createResponseMessage();
-  JAST &error = errorMessage.add("error", JSON_OBJECT);
-  error.add("code", JSON_INTEGER, code);
-  error.add("message", message.c_str());
-  sendMessage(errorMessage);
-}
+        // Log the request
+        clientLog << content << std::endl;
 
-void sendErrorMessage(JAST receivedMessage, const char *code, std::string message) {
-  JAST errorMessage = createResponseMessage(receivedMessage);
-  JAST &error = errorMessage.add("error", JSON_OBJECT);
-  error.add("code", JSON_INTEGER, code);
-  error.add("message", message.c_str());
-  sendMessage(errorMessage);
-}
+        // Parse that content as JSON
+        JAST request;
+        std::stringstream parseErrors;
+        if (!JAST::parse(content, parseErrors, request)) {
+          sendErrorMessage(ParseError, parseErrors.str());
+        } else {
+          const std::string &method = request.get("method").value;
+          if (!isInitialized && (method != "initialize")) {
+            sendErrorMessage(request, ServerNotInitialized, "Must request initialize first");
+          } else if (isShutDown && (method != "exit")) {
+            sendErrorMessage(request, InvalidRequest, "Received a request other than 'exit' after a shutdown request.");
+          } else {
+            callMethod(method, request);
+          }
+        }
+      }
+    }
 
+    void callMethod(std::string method, JAST request) {
+      auto functionPointer = methodToFunction.find(method);
+      if (functionPointer != methodToFunction.end()) {
+          ((*this).*(functionPointer->second))(request);
```

Comment:
Does `this->*(functionPointer->second)(request)` work?

Reply:
Yes, thank you